From 5664a16b1b05441c471637215d2dd9868c10d80a Mon Sep 17 00:00:00 2001
From: Thomas Bourgoin <thomas.bourgoin@foss.st.com>
Date: Mon, 26 Feb 2024 14:18:29 +0100
Subject: [PATCH 0670/1695] crypto: stm32: PKA limit driver support to ECDSA
 and ECDH.

Fix regression of xtest 4006 testing TEE_ALG_SM2_PKE.
PKA only support ECDSA key generation ECDH key exchange.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git bce68bf91574d1c7aedd2d309ee852fd9382ad33 crypto: stm32: PKA limit driver support to ECDSA and ECDH.]
Change-Id: I2a2fdbc0fdd20d89bbf9ea9c4604df818fcaf54e
Signed-off-by: Thomas Bourgoin <thomas.bourgoin@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/362215
---
 core/drivers/crypto/stm32/ecc.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/core/drivers/crypto/stm32/ecc.c b/core/drivers/crypto/stm32/ecc.c
index f8d585c25..02c9c55a0 100644
--- a/core/drivers/crypto/stm32/ecc.c
+++ b/core/drivers/crypto/stm32/ecc.c
@@ -294,6 +294,11 @@ static TEE_Result stm32_alloc_keypair(struct ecc_keypair *s, uint32_t type,
 	if (!s)
 		return TEE_ERROR_BAD_PARAMETERS;
 
+	/* This driver only supports ECDH/ECDSA */
+	if (type != TEE_TYPE_ECDSA_KEYPAIR &&
+	    type != TEE_TYPE_ECDH_KEYPAIR)
+		return TEE_ERROR_NOT_IMPLEMENTED;
+
 	DMSG("Using PKA");
 	memset(s, 0, sizeof(*s));
 
@@ -323,6 +328,11 @@ static TEE_Result stm32_alloc_publickey(struct ecc_public_key *s, uint32_t type,
 	if (!s)
 		return TEE_ERROR_BAD_PARAMETERS;
 
+	/* This driver only supports ECDH/ECDSA */
+	if (type != TEE_TYPE_ECDSA_KEYPAIR &&
+	    type != TEE_TYPE_ECDH_KEYPAIR)
+		return TEE_ERROR_NOT_IMPLEMENTED;
+
 	DMSG("Using PKA");
 	memset(s, 0, sizeof(*s));
 
-- 
2.47.2

